### PR TITLE
perf(CharZero): de-simp `OfNat.ofNat_eq_ofNat` 

### PR DESCRIPTION
**Pull request by mattrobball**
This is attempted 46478 times for `simp` and is successful 0 times.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mattrobball** [comment]
!bench

**Comment by MichaelStollBayreuth** [comment]
<b>General information:</b>
```text
lint:                                                  -9.6715 * 10⁹ (-0.138%)
build:                                                 -68.003 * 10⁹ (-0.0552%)
```


<b>2 files got slower by at least 10⁹ instructions:</b>
```text
Mathlib.NumberTheory.RamificationInertia:              +5.4142 * 10⁹ (+1.79%)
Mathlib.Algebra.Exact:                                 +4.4683 * 10⁹ (+2.99%)
```


<b>No file got slower by at least 10%.</b>


<b>10 files got faster by at least 10⁹ instructions:</b>
```text
Mathlib.Algebra.Homology.DifferentialObject:           -3.1134 * 10⁹ (-2.02%)
Mathlib.CategoryTheory.Triangulated.Functor:           -2.6756 * 10⁹ (-0.777%)
Mathlib.GroupTheory.Coxeter.Matrix:                    -2.5101 * 10⁹ (-2.39%)
Mathlib.Algebra.Homology.HomotopyCategory.SingleFunctors:
                                                       -1.8658 * 10⁹ (-2.17%)
Mathlib.AlgebraicTopology.SimplicialObject:            -1.3443 * 10⁹ (-0.640%)
Mathlib.CategoryTheory.Monoidal.Internal.Module:       -1.2882 * 10⁹ (-0.730%)
Mathlib.Topology.Sheaves.SheafCondition.OpensLeCover:  -1.2523 * 10⁹ (-1.31%)
Mathlib.LinearAlgebra.Multilinear.Basic:               -1.2491 * 10⁹ (-0.842%)
Mathlib.CategoryTheory.Idempotents.HomologicalComplex: -1.0946 * 10⁹ (-2.54%)
Mathlib.RepresentationTheory.Action.Limits:            -1.0013 * 10⁹ (-1.47%)
```


<b>No file got faster by at least 10%.</b>

**Comment by MichaelStollBayreuth** [comment]
There is a net positive, but it is a bit strange that two files got a bit slower. (Not sure if this is within the "noise" regime...)

**Comment by mattrobball** [comment]
Probably noise. Removing something rejiggers simp priorities slightly so maybe that flaps its wings and ... 

It is worth noting there are 100+ more of these non-applying simp theorems with more than 100 uses. Some have some pretty big type classes assumptions. I am interested to see the effect of bulk removal

**Comment by eric-wieser** [comment]
> This is attempted 46478 times for `simp` and is successful 0 times.

The fact a lemma is never used by mathlib does not mean it is not a good simp lemma; we should consider downstream projects too.

That being said, this particular case is probably more in scope for norm_num. Can you add a test that norm_num can solve this, and a comment on the lemma explaining that it is not `simp` as `norm_num` can handle it?

**Comment by mattrobball** [comment]
I think that being attempted in rewrite almost 50000 times and failing every single time is the definition of a bad simp theorem. 

I'll check norm_num.

**Comment by eric-wieser** [comment]
> I think that being attempted in rewrite almost 50000 times and failing every single time is the definition of a bad simp theorem.

I think the true blame here falls on `OfNat` not being used in DiscrTrees, but that requires a change in core.
A comment referencing the lean4 issue would be fine instead, if the `norm_num` comment is false.

File: Mathlib/Algebra/CharZero/Defs.lean [hunk 1]
```diff
@@ -93,6 +93,9 @@ end Nat
 
 namespace OfNat
 
+-- TODO: This whole section is draped in `no_index` because of its special casing in
+-- core, see leanprover/lean4#2867. Hopefully this can be fixed in the future.
+
```
Comment (eric-wieser) [inline Mathlib/Algebra/CharZero/Defs.lean:98]: The standard comment we use (on every lemma) is `-- See note [no_index around OfNat.ofNat]`

**Pull request closed without merging**